source/engine/graphics/system/SkeletalAnimationCalculations.cpp

```diff
@@ -0,0 +1,276 @@
+    animatedBones.reserve(rig.vertexToBone.size());
+
+    // Create a final transform for each bone by multiplying the (vertex-space-to-bone-space matrix) with the (bone-space-to-animated-parent-bone-space matrix) with the (global inverse transform matrix).
+    // This outputs a matrix that can be used to transform a vertex into its final animated position.
+    auto globalInverseTransform = rig.globalInverseTransform;
+    auto vertexToBoneSize = rig.vertexToBone.size();
+    for (auto i = 0u; i < vertexToBoneSize; i++)
+    {
+        animatedBones.push_back(nc::graphics::SkeletalAnimationData{rig.vertexToBone[i] * boneToParentSandbox[rig.offsetsMap[i]] * globalInverseTransform});
+    }
+
+    return animatedBones;
+}
+
+auto GetInterpolatedPosition(float timeInTicks, const std::vector<nc::asset::PositionFrame> positionFrames) -> nc::Vector3
+{
+    auto interpolatedPosition = nc::Vector3{};
+
+    if (positionFrames.size() == 0)
+    {
+        throw NcError("Animation has no position data for the node.");
+    }
+
+    if (positionFrames.size() == 1)
+    {
+        return positionFrames[0].position;
+    }
+
+    auto positionIndex = UINT_LEAST32_MAX;
+    for (auto i = 0u; i < positionFrames.size(); i++)
+    {
+        if (timeInTicks < positionFrames[i+1].timeInTicks)
+        {
+            positionIndex = i;
+            break;
+        }
+    }
+    if (positionIndex == UINT_LEAST32_MAX)
+    {
+        throw NcError("Animation has no position data at time: ", std::to_string(timeInTicks));
+    }
+
+    auto nextPositionIndex = positionIndex + 1;
+    if ((positionIndex + 1) > positionFrames.size())
+    {
+        throw NcError("Animation interpolation index has exceeded the bounds of the frame data.");
+    }
+
+    auto deltaTimeInTicks = positionFrames[nextPositionIndex].timeInTicks - positionFrames[positionIndex].timeInTicks;
+    auto interpolationFactor = (timeInTicks - positionFrames[positionIndex].timeInTicks) / deltaTimeInTicks;
+
+    if (interpolationFactor < 0.0f || interpolationFactor > 1.0f)
+    {
+        throw NcError("Error calculating the interpolation factor: ", std::to_string(interpolationFactor));
+    }
+
+    interpolatedPosition = Lerp(positionFrames[positionIndex].position, positionFrames[nextPositionIndex].position, static_cast<float>(interpolationFactor));
+    return interpolatedPosition;
+}
+
+auto GetInterpolatedRotation(float timeInTicks, const std::vector<nc::asset::RotationFrame> rotationFrames) -> nc::Quaternion
+{
+    auto interpolatedRotation = nc::Quaternion{};
+
+    if (rotationFrames.size() == 0)
+    {
+        throw NcError("Animation has no rotation data.");
+    }
+
+    if (rotationFrames.size() == 1)
+    {
+        return Normalize(rotationFrames[0].rotation);
+    }
+
+    auto rotationIndex = UINT_LEAST32_MAX;
+    for (auto i = 0u; i < rotationFrames.size(); i++)
+    {
+        if (timeInTicks < rotationFrames[i+1].timeInTicks)
```

Comment:
Hmm, I already typed this comment, but it disappeared. Apologies if it duplicates.

i+1 goes out of bounds on the last iteration. Maybe use a `std::find_if` and something along the lines of `std::distance(beg, std::prev(pos))` to get the index. Probably requires a check to make sure we don't decrement from begin.

Reply:
Ok, I should have look a bit closer. Based on code below, it looks like it *shouldn't* read past the end, but it still makes me feel icky to have to potential looming around. I don't think you need the index at all; just dereference the iterator(s) instead of doing a subscript.